### PR TITLE
GovernanceV4- routes simulation

### PR DESCRIPTION
**Pull request by 0xlucian**
## Description

This PR adds forked mainnet tests for the new GovernanceV4 upgrade. Along with the tests, it introduces a new approach of creating forked mainnet tests by calling `setForkBlock(blockNumber);` in order to fork a chain from a certain block. This allows us to specify the fork block more glanularly and is a bit more flexible.

## Checklist

- [ ] I have updated the documentation to account for the changes in the code.
- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [ ] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: contracts/Governance/Timelock.sol [hunk 1]
```diff
@@ -34,8 +34,8 @@ contract Timelock {
     );
 
     uint public constant GRACE_PERIOD = 14 days;
-    uint public constant MINIMUM_DELAY = 2 days;
-    uint public constant MAXIMUM_DELAY = 30 days;
+    uint public constant MINIMUM_DELAY = 800; // About 1 hour, 3 secs per block (20 * 60)
+    uint public constant MAXIMUM_DELAY = 28800; // About 24 hours, 3 secs per block (20 * 60 * 24)
```
Comment (kkirka) [inline contracts/Governance/Timelock.sol:38]: But the delay is in seconds here

Comment (kkirka) [reply contracts/Governance/Timelock.sol:38]: I propose we keep the upper bound > 2 days because our current timelock has a delay of 2 days

File: script/hardhat/fork/govUpdateTest.ts [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+    .initialize(
+      XVS_VAULT_MAINNET,
+      PROPOSAL_TYPE_CONFIGS,
+      [NORMAL_VIP_TIMELOCK, timeLockFastTrack.address, timeLockCritical.address],
+      await governorAdmin.getAddress(),
+    );
+  await tx.wait();
+};
+
+describe("Governcance v4", () => {
```
Comment (kkirka) [inline script/hardhat/fork/govUpdateTest.ts:114]:
```suggestion
describe("Governance v4", () => {
```

File: contracts/Governance/Timelock.sol (hunk 1 above)
Comment (kkirka) [reply contracts/Governance/Timelock.sol:38]: I'm 100% sure it's in seconds: https://github.com/VenusProtocol/venus-protocol/blob/develop/contracts/Governance/Timelock.sol#L89

File: script/hardhat/fork/govUpdateTest.ts [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+        const proposalIdBefore = await governorProxy.callStatic.proposalCount();
+        const tx = await governorProxy
+          .connect(proposer)
+          .propose(targets, values, signatures, getCalldatas(proposal), "VIP 1", PROPOSAL_TYPES.NORMAL);
+        await tx.wait();
+        proposalId = await governorProxy.callStatic.proposalCount();
+        expect(proposalIdBefore.add(1)).to.equal(proposalId);
+      });
+    });
+    describe("Cast Vote:", () => {
```
Comment (coreyar) [inline script/hardhat/fork/govUpdateTest.ts:145]: Can we add some spacing in between test suites and even inside test cases for readability? At minimum we should have a new line between suites and cases

File: script/hardhat/fork/govUpdateTest.ts [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+let supporter: SignerWithAddress;
+let governorAdmin: SignerWithAddress;
+let aclAdmin: SignerWithAddress;
+let timeLockFastTrack: TimelockInterface;
+let timeLockCritical: TimelockInterface;
+
+const initMainnetUser = async (user: string, balance: number) => {
+  await impersonateAccount(user);
+  await setBalance(user, balance);
+  return ethers.getSigner(user);
```
Comment (coreyar) [inline script/hardhat/fork/govUpdateTest.ts:66]: if the goal is to just return the signer. We should we un-impersonate after. Otherwise the connected account remains the last impersonated

File: script/hardhat/fork/govUpdateTest.ts [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+    };
+    let proposalId;
+    before(async () => {
+      await loadFixture(governanceFixture);
+    });
+    describe("Propose:", () => {
+      it("should revert if below votingThreshold", async () => {
+        const { targets, signatures, values } = proposal;
+        await expect(
+          governorProxy.propose(targets, values, signatures, getCalldatas(proposal), "VIP 1", PROPOSAL_TYPES.NORMAL),
```
Comment (coreyar) [inline script/hardhat/fork/govUpdateTest.ts:131]: nit: This would be a little clearer if we labeled who was calling. Here we assume we are connected to the default account and that doesn't have enough XVS staked. This isn't exactly true though because we are using impersonate account without stopping the impersonation

File: script/hardhat/fork/govUpdateTest.ts [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+  });
+  describe("FASTTRACK VIP", () => {
+    const proposal = {
+      signatures: ["_setLiquidationIncentive(uint256)"],
+      targets: [COMPTROLLER_PROXY_MAINNET],
+      values: ["0"],
+      params: [[NEW_LIQ_INCENTIVE]],
+    };
+    let proposalId;
+    before(async () => {
```
Comment (coreyar) [inline script/hardhat/fork/govUpdateTest.ts:194]: space around before blocks

Comment (coreyar) [reply script/hardhat/fork/govUpdateTest.ts:194]: maybe there is a linting rule for this

File: script/hardhat/fork/govUpdateTest.ts [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+
+const TIMELOCK_DELAYS_MAINNET = {
+  NORMAL: 172800,
+  FAST_TRACK: SIX_HOURS,
+  CRITICAL: ONE_HOUR,
+};
+
+const PROPOSAL_TYPES = {
+  NORMAL: 0,
+  FASTTRACK: 1,
```
Comment (coreyar) [inline script/hardhat/fork/govUpdateTest.ts:49]: FAST_TRACK to match solidity enum

Comment (coreyar) [reply script/hardhat/fork/govUpdateTest.ts:49]: hmm actually the solidity enum is `FASTTRACK` but since it is two words it should be `FAST_TRACK`

File: script/hardhat/fork/govUpdateTest.ts [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+          .withArgs(await supporter.getAddress(), proposalId, 1, 846441499442385723656125n, "");
+      });
+    });
+    describe("Queue Proposal:", () => {
+      it("should revert if voting period is not passed", async () => {
+        await expect(governorProxy.connect(proposer).queue(proposalId)).to.be.revertedWith(
+          "GovernorBravo::queue: proposal can only be queued if it is succeeded",
+        );
+      });
+      it("should be queued sucessfully", async () => {
```
Comment (coreyar) [inline script/hardhat/fork/govUpdateTest.ts:233]:
```suggestion
      it("should be queued successfully", async () => {
```

File: script/hardhat/fork/govUpdateTest.ts [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+        await expect(governorProxy.connect(proposer).castVote(proposalId, 1))
+          .to.emit(governorProxy, "VoteCast")
+          .withArgs(await proposer.getAddress(), proposalId, 1, 335338539326609831760663n, "");
+
+        await expect(governorProxy.connect(supporter).castVote(proposalId, 1))
+          .to.emit(governorProxy, "VoteCast")
+          .withArgs(await supporter.getAddress(), proposalId, 1, 846441499442385723656125n, "");
+      });
+    });
+    describe("Queue Proposal:", () => {
```
Comment (coreyar) [inline script/hardhat/fork/govUpdateTest.ts:227]:
```suggestion
    describe("Queue Proposal", () => {
```
I am fairly certain that hardhat will add necessary punctuation when rendering test output

File: script/hardhat/fork/govUpdateTest.ts [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+      it("should revert if voting period is not passed", async () => {
+        await expect(governorProxy.connect(proposer).queue(proposalId)).to.be.revertedWith(
+          "GovernorBravo::queue: proposal can only be queued if it is succeeded",
+        );
+      });
+      it("should be queued sucessfully", async () => {
+        await mineUpTo((await ethers.provider.getBlockNumber()) + PROPOSAL_TYPE_CONFIGS[0].votingPeriod);
+        const tx = await governorProxy.connect(proposer).queue(proposalId);
+        await tx.wait();
```
Comment (coreyar) [inline script/hardhat/fork/govUpdateTest.ts:166]: Should we have an assertion here or are we just checking it doesn't revert?

File: script/hardhat/fork/govUpdateTest.ts [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+    describe("Execute Proposal:", () => {
+      it("should revert if timelock period has not passed", async () => {
+        await expect(governorProxy.connect(proposer).execute(proposalId)).to.be.revertedWith(
+          "Timelock::executeTransaction: Transaction hasn't surpassed time lock.",
+        );
+      });
+      it("should be executed successfully", async () => {
+        await mineUpTo((await ethers.provider.getBlockNumber()) + TIMELOCK_DELAYS_MAINNET.NORMAL);
+        const tx = await governorProxy.connect(proposer).execute(proposalId);
+        await tx.wait();
```
Comment (coreyar) [inline script/hardhat/fork/govUpdateTest.ts:178]: Should we have an assertion here or are we just checking it doesn't revert?

File: script/hardhat/fork/govUpdateTest.ts [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+      });
+    });
+    describe("Execute Proposal:", () => {
+      it("should revert if timelock period has not passed", async () => {
+        await expect(governorProxy.connect(proposer).execute(proposalId)).to.be.revertedWith(
+          "Timelock::executeTransaction: Transaction hasn't surpassed time lock.",
+        );
+      });
+      it("should be executed successfully", async () => {
+        await mineUpTo((await ethers.provider.getBlockNumber()) + TIMELOCK_DELAYS_MAINNET.FAST_TRACK);
```
Comment (coreyar) [inline script/hardhat/fork/govUpdateTest.ts:246]: Here we say `FAST_TRACK` we should be consistent although this is more correct

File: script/hardhat/fork/govUpdateTest.ts [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+        const tx = await governorProxy.connect(proposer).execute(proposalId);
```
Comment (coreyar) [inline script/hardhat/fork/govUpdateTest.ts:247]: missing assertion

File: script/hardhat/fork/utils.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+      // Ensure no inline comments are parsed and trim the whitespace.
+      return arg.replace(/\/\*.*\*\//, "").trim();
+    })
+    .filter(arg => {
+      // Ensure no undefined values are added.
+      return arg;
+    });
+};
+
+module.exports = {
```
Comment (coreyar) [inline script/hardhat/fork/utils.ts:42]: This is a typescript file so it uses ejs there form we should use the `export` keyword instead of the commonjs module system

**Review by kkirka — APPROVED** [review]

**Review by defcon022 — APPROVED** [review]
LGTM
